src/secureSignal_shared.ts

```diff
@@ -0,0 +1,108 @@
+const MAXIMUM_RETRY = 3;
+export class UidSecureSignalProvider {
+  debug: boolean;
+
+  constructor(debug = false) {
+    this.debug = debug;
+
+    if (typeof window.getUidAdvertisingToken === 'function') {
+      this.logging('register SecureSignalProvider');
+      this.registerSecureSignalProvider();
+    }
+  }
+
+  public registerSecureSignalProvider = () => {
+    const uidHandler = this.retrieveAdvertisingTokenHandler();
+
+    if (!uidHandler) {
+      console.warn('UidSecureSignal: Please implement `getUidAdvertisingToken`');
+      return;
+    }
+
+    window.googletag = window.googletag || {
+      cmd: [],
+    };
+
+    window.googletag.secureSignalProviders = window.googletag.secureSignalProviders || [];
+    window.googletag.secureSignalProviders.push({
+      id: 'uidapi.com',
+      collectorFunction: async () => {
+        this.logging('collectorFunction invoked');
+        const uidAdvertisingToken = await getUidAdvertisingTokenWithRetry(uidHandler);
+        this.logging(`collectorFunction pushes: ${uidAdvertisingToken}`);
+        return uidAdvertisingToken;
+      },
+    });
+  };
+
+  public logging = (message: string) => {
+    if (!this.debug) return;
+    console.log(`UidSecureSignal: ${message}`);
+  };
+
+  private retrieveAdvertisingTokenHandler = (): Function | undefined => {
+    if (typeof window.getUidAdvertisingToken === 'function') {
+      return window.getUidAdvertisingToken!;
+    }
+
+    if (window.__uid2 && 'getAdvertisingTokenAsync' in window.__uid2!) {
```

Comment:
Just check for whichever SDK is relevant to this provider.

Reply:
So I used a flag when initializing the provider but I'm wondering if you know a better way.